### ISSUE TITLE
Report model update issues due to concurrent edits separately?

### ISSUE THREAD

**Issue by enikao** [LionWeb-io/specification#238]
Assume client A retrieved some partition at time 1.

Example nodes with containment:
```
P
  X
    Y
  Z
```

Then, client B updated that partition at time 2 (move `Z` from `P` to `Y`)
```
  Y
    Z
```

At time 3, client A stores an update to that partition (move `X` from `P` to `Z`). The update would be valid from its perspective, but illegal after B's update: It would introduce a circle in the dependency tree.

```
  Z
    X
```

The repository cannot allow that. How does it report back?

## Variant A: Just report "tree shape error"
Same as if client A sent an update that would violate the tree shape even from client A's perspective (most probably a programming bug in client A).

Pro:
* Simple to implement

Con:
* Client can't differentiate between programming bugs and concurrent edits.

## Variant B: Report "concurrency error"
The repository informs the client that the error is not based on programming issues, but on concurrent edits.

For this, the repository must have some means to detect whether the client is using the latest contents of the repository.
There are different ways to do this.

### Variant B1: Repository tracks internally the state of each client
If the repository can track each client's "last seen" state, it might assume or analyze the model update error is based on concurrent edits.

Pro:
* No client support needed

Con:
* Might be hard to implement for the repository
* Repository has to keep track of the latest state for each client

### Variant B2: Client informs repository about last seen state
Every time the repository replies with some part of its contents, it sends along an [ETag](https://en.wikipedia.org/wiki/HTTP_ETag) representing the state of the repository's contents.
The client sends the latest ETag back with the next request. Then the repository can compare its current ETag with the requested ETag and deduct a concurrency issue from there.

Pro:
* Scales nicely with large amount of clients

Con:
* It might be hard for a repository to keep its global state. It could use an internal counter and increase with every change, but then needs to store that counter somewhere. It might also use a "hash" over the complete content, but that might be expensive to calculate.

**Comment by enikao** [comment]
## Decision: Optional variant B1: Report concurrent issues if possible

A repository SHOULD report concurrent edits if it can; otherwise, it MUST report a generic tree shape error.

We will define a dedicated _message kind_.
We will not include any ETag support in Bulk API.

**Comment by joswarmer** [comment]
I wasn't at the discussion and have some questions

- What is a "generic tree shape error"? That term is new to me.
- Currently the the repository reports errors _as specifically as possible_ using the _Message kind_ defined in the bulk api spec. Any tree error is (supposed to be) one of the defined message kinds.

How do these two relate to each other?

**Comment by joswarmer** [comment]
I would strongly prefer to make this OPTIONAL instead of SHOULD, because of the complexity of implementing this.

**Comment by enikao** [comment]
> What is a "generic tree shape error"? That term is new to me.

Yes, it's not a term we've used before. It means a message that indicates that the nodes violate one of our tree constraints (mismatching parent / containment, more than one containment for the same node, etc.). Whether we report that with one error kind or several is TBD.

>  Currently the the repository reports errors _as specifically as possible_ using the _Message kind_ defined in the bulk api spec. Any tree error is (supposed to be) one of the defined message kinds.

I also tend toward specific messages.

**Comment by enikao** [comment]
I think this issue conflicts with #297

**Comment by enikao** [comment]
> I think this issue conflicts with [#297](https://github.com/LionWeb-io/specification/issues/297)

The repo might still know about the conflict if it received both commands, but didn't send out any events yet.
